Drag & drop items from Library UI

aosyatnik opened the pull request:
Drag an entry in the Library UI, and drop it on the canvas, then a node is created.

Added new class `DragDropNodeSearchElementInfo`. It's used in `OnWorkspaceDrop`, so that we can get type of all inherited classes(e.g. ZeroTouchNodeSearchElement, CustomNodeSearchElement etc.).

Reviewers
@Benglin , please take a look.

Link to YouTrack:
[MAGN-6060](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-6060) DUI: Make items drag-able from library

pboyer commented on src/DynamoCoreWpf/Controls/StandardPanel.xaml.cs:
```diff
@@ -47,7 +47,18 @@ private void OnHeaderButtonClick(object sender, RoutedEventArgs e)
+
+            var listBoxItem = sender as ListBoxItem;
+            var searchElementVM = listBoxItem.DataContext as NodeSearchElementViewModel;
+
```
@aosyatnik We should make sure to check these two properties for null here.  If they are null, we should return prematurely.

pboyer commented:
@Benglin This looks very reasonable right now.  What do you think?

Benglin commented on src/DynamoCoreWpf/Views/Core/WorkspaceView.xaml.cs:
```diff
@@ -702,6 +703,18 @@ private HitTestResultBehavior DirectHitTestCallback(HitTestResult result)
+            var nodeModel = nodeInfo.SearchElement.CreateNode();
+            var mousePosition = e.GetPosition(this.WorkspaceElements);
+            ViewModel.DynamoViewModel.
+                ExecuteCommand(new DynamoModel.CreateNodeCommand(nodeModel, mousePosition.X, mousePosition.Y, false, true));
```
Perhaps a better line-wrapping would be like so:

```csharp
var nodeModel = nodeInfo.SearchElement.CreateNode();
var mousePosition = e.GetPosition(this.WorkspaceElements);
ViewModel.DynamoViewModel.ExecuteCommand(new DynamoModel.CreateNodeCommand(
    nodeModel, mousePosition.X, mousePosition.Y, false, true));
```

Benglin commented on src/DynamoCoreWpf/Controls/StandardPanel.xaml.cs:
```diff
@@ -47,7 +47,18 @@ private void OnHeaderButtonClick(object sender, RoutedEventArgs e)
+            DragDrop.DoDragDrop(listBoxItem, new DragDropNodeSearchElementInfo(searchElementVM.Model), DragDropEffects.Copy);
+        }
+
+        private void OnMouseLeftButtonUp(object sender, MouseButtonEventArgs e)
```
Great! Having `up` create a node is better than `down` (just to align with button presses throughout the system).

Benglin commented:
Thanks for the help @pboyer! Hi @aosyatnik, minor tweaks and we're good to go! :)

aosyatnik commented:
Hi @pboyer and @Benglin , thank you for review. I changed thin places. But after some tests I found, that not everything works. If I tried to drag items, that aren't parts of class(e.g. builtin, operators), drag&drop didn't work, since they were not parts of `StandardPanel`.
So, I have to move my `DoDragDrop` in the above Layer(`LibraryView`). There was 1 problem: In `LibraryView` all items are presented as Buttons. That's why we can't use PreviewUp and PreviewDown events as in list items. 
I have to detect whether user is dragging or clicking button.
@Benglin , please look through, thank you.

pboyer commented:
Hi @aosyatnik,

I pulled this, built it, and gave it a try.  Here are the main issue I noticed:  I cannot drag and drop items from the library after searching (e.g. type "number", then try and drag and drop first item)

If we fix that, I think this is pretty much ready to go.

aosyatnik commented:
Hi, @pboyer , @Benglin .
I didn't know, that search has to have drag & drop functionality. Added it.

pboyer commented:
@aosyatnik No problem!  @Benglin, what do you think?

Benglin commented:
Thanks a lot for helping, @pboyer! You raised all valid points, appreciate that.

Hi @aosyatnik, there is a problem with dragging `Operator` and `Custom Node` nodes from the Library UI. It appears that `OnExpanderButtonPreviewMouseMove` is called twice, thus creating two nodes at once (this behavior is harder to see in the debugger due to the lost of focus, but you can define a static variable `a` and then break conditionally when `a % 2 == 0`):

```
DynamoCoreWpf.dll!Dynamo.UI.Views.LibraryView.StartDrag(System.Windows.Input.Mo...
DynamoCoreWpf.dll!Dynamo.UI.Views.LibraryView.OnExpanderButtonPreviewMouseMove(...
[External Code]
DynamoCoreWpf.dll!Dynamo.UI.Views.LibraryView.StartDrag(System.Windows.Input.Mo...
DynamoCoreWpf.dll!Dynamo.UI.Views.LibraryView.OnExpanderButtonPreviewMouseMove(...
[External Code]
DynamoSandbox.exe!DynamoSandbox.Program.MakeStandaloneAndRun(string commandFile...
DynamoSandbox.exe!DynamoSandbox.Program.Main(string[] args) Line 186    C#     ...
```

Once that's fixed this can be merged (also those points that Peter raised). Thanks!

aosyatnik commented:
@Benglin , thank you for finding this bug. I fixed it by adding `IsDragging` property.
@pboyer , thank you for review, I've updated cursor... I'm not sure, that I got the idea right, please take a look.

pboyer commented:
Hi Anna,

I've created a follow up task for changing the icon:  http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-7377

For now, this looks good to me.  Awaiting the final go-ahead from @Benglin.  :)

pboyer commented:
Hi @aosyatnik,

This looks solid to me and I think it would be great to get it in front of users and tested.  I'm merging it in.

@Benglin,

If you have any remaining issues with this, can we open up separate follow up tasks?

Benglin commented:
Thanks @pboyer, since that drag-and-drop issue has been fixed, I'm good with you merging this in. I was completely occupied by localization build stuff yesterday so I didn't have a chance with this. Thanks again for your help, and also for @aosyatnik's great work :)

Benglin commented on src/DynamoCoreWpf/Views/Search/LibraryView.xaml.cs:
```diff
@@ -285,5 +286,67 @@ private void OnAddButtonPreviewMouseDown(object sender, MouseButtonEventArgs e)
+                return;
+            }
+
+            DragDrop.DoDragDrop(senderButton, new DragDropNodeSearchElementInfo(searchElementVM.Model), DragDropEffects.Copy);
```
Should we only set `IsDragging = true` right before `DragDrop.DoDragDrop`? Or even better, can we do it only inside `DragDrop.DoDragDrop` method so it becomes self-contained?
